daemon/emer-circular-file.c

```diff
@@ -0,0 +1,825 @@
+
+  object_class->set_property = emer_circular_file_set_property;
+  object_class->finalize = emer_circular_file_finalize;
+
+  emer_circular_file_props[PROP_PATH] =
+    g_param_spec_string ("path", "Path",
+                         "Path at which circular file is stored",
+                         NULL,
+                         G_PARAM_CONSTRUCT_ONLY | G_PARAM_WRITABLE |
+                         G_PARAM_STATIC_STRINGS);
+
+  emer_circular_file_props[PROP_MAX_SIZE] =
+    g_param_spec_uint64 ("max-size", "Max size",
+                         "The maximum permitted physical size of the "
+                         "underlying data file. Does not include the overhead "
+                         "of the metadata file.",
+                         0, G_MAXUINT64, 0,
+                         G_PARAM_CONSTRUCT_ONLY | G_PARAM_WRITABLE |
+                         G_PARAM_STATIC_STRINGS);
+
+  g_object_class_install_properties (object_class, NPROPS,
+                                     emer_circular_file_props);
+}
+
+static void
+emer_circular_file_init (EmerCircularFile *self)
+{
+  EmerCircularFilePrivate *priv =
+    emer_circular_file_get_instance_private (self);
+
+  priv->write_buffer = g_byte_array_new ();
+}
+
+static gboolean
+emer_circular_file_may_fail_init (GInitable    *initable,
+                                  GCancellable *cancellable,
+                                  GError      **error)
+{
+  EmerCircularFile *self = EMER_CIRCULAR_FILE (initable);
+  EmerCircularFilePrivate *priv =
+    emer_circular_file_get_instance_private (self);
+
+  gchar *metadata_filepath = get_metadata_filepath (self);
+  priv->metadata_key_file = g_key_file_new ();
+  GError *local_error = NULL;
+  gboolean load_succeeded =
+    g_key_file_load_from_file (priv->metadata_key_file, metadata_filepath,
+                               G_KEY_FILE_NONE, &local_error);
+  g_free (metadata_filepath);
+  if (!load_succeeded)
+    {
+      if (!g_error_matches (local_error, G_KEY_FILE_ERROR,
+                            G_KEY_FILE_ERROR_NOT_FOUND) &&
+          !g_error_matches (local_error, G_FILE_ERROR, G_FILE_ERROR_NOENT))
+        {
+          g_propagate_error (error, local_error);
+          return FALSE;
+        }
+
+      guint64 little_endian_max_size =
+        swap_bytes_64_if_big_endian (priv->max_size);
+      g_key_file_set_uint64 (priv->metadata_key_file, METADATA_GROUP_NAME,
+                             MAX_SIZE_KEY, little_endian_max_size);
+      return set_metadata (self, 0, 0, error);
+    }
+
+  guint64 little_endian_size =
+    g_key_file_get_uint64 (priv->metadata_key_file, METADATA_GROUP_NAME,
+                           SIZE_KEY, &local_error);
+  if (local_error != NULL)
+    goto handle_failed_read;
+
+  priv->size = swap_bytes_64_if_big_endian (little_endian_size);
+  if (priv->size > priv->max_size)
+    {
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_INVALID_DATA, "Logical size "
```

Comment:
Could also set the local error here and go to the common cleanup function for consistency.